### ISSUE TITLE
✅ 🐞 CSS Shorthand for Gap Values Makes Values Unusable with CSS

### ISSUE THREAD

**Issue by KingGrizzly** [elementor/elementor#24178]
### Prerequisites

- [X] I have searched for similar issues in both open and closed tickets and cannot find a duplicate.
- [X] The issue still exists against the latest stable version of Elementor.

### Description

In Elementor 3.15.x, the global and container gap setting was just one value. For example, `gap: 12px`.
The CSS variable elementor created would look like:

```css
elementor-element {
--gap: 12px;
}
```

In Elementor 3.16.x there is a column and row gap value (See PR: https://github.com/elementor/elementor/pull/21672). When this is outputted to CSS, it is rendered in CSS shorthand:

```css
elementor-element {
--gap: 18px 12px;
}
```

This has broken at least 6 client sites (we are facing weeks of manual adjustment if this is not addressed) because we were using variables for the gap settings, for example: --gap-300p where the associated CSS could use calc such as:

`calc((--gap) * 3)`
Effectively, `12px * 3 = 36px`

This would allow us to use shorthand variables in our containers. `--gap-400p`, `--gap-50p`, etc.

Since Elementor has gone to two container gap values but writes the CSS using Flexbox gap shorthand, this breaks variable support. For example, if the column gap was 12px and the row gap was 24px, the calc does something like:

`calc((12px 24px) *3)`

The CSS calc doesn't work, breaking the layouts of every page. I think Elementor needs to keep the two flex box gap variables for column and row (see: https://css-tricks.com/almanac/properties/g/gap/ ), and possibly keep the shorthand so nothing breaks:

```css
--gap
--gap-column
--gap-row
```

Or, does anyone know how to pick out those two values separately? Here's a video explaining the use case for CSS variables in the custom value field: https://www.loom.com/share/0a3b59844d59438cb9719ccfdcf32bea?sid=7bed4478-176f-442d-8b98-e71298899c07

### Steps to reproduce

- Set gap values globally (under site settings) or on a parent container. Try 24px row / 12 px column
- Place a child container in the parent container and give it a class (.child-test)
- Set the gap values to variables on the child: --gap-300p / --gap-50p

Write some CSS for the child:

```css
.child-test {
--gap-300p: calc((--gap) * 3);
--gap-50p: calc((--gap) * .5);
}
```

Study the results on the front end using console.

Notice how the variable calc CSS can't work because you'll get a result something like:

`calc((24px 12px) *3)`

We need a way to fetch the column and row gap values separately.

### Isolating the problem

- [X] This bug happens with only Elementor plugin active (and Elementor Pro).
- [X] This bug happens with a Blank WordPress theme active ([Hello theme](https://wordpress.org/themes/hello-elementor/)).
- [X] I can reproduce this bug consistently following the steps above.

### System Info

<details>
<summary>Click to reveal</summary>

````txt

== Server Environment ==
Operating System: Linux
Software: nginx/1.25.2
MySQL version: MariaDB Server v10.4.25
PHP Version: 8.1.24
PHP Memory Limit: 256M
PHP Max Input Vars: 10000
PHP Max Post Size: 100M
GD Installed: Yes
ZIP Installed: Yes
Write Permissions: All right
Elementor Library: Connected

== WordPress Environment ==
Version: 6.3.2
Site URL: https://mysite.com/
Home URL: https://mysite.com/
WP Multisite: No
Max Upload Size: 100 MB
Memory limit: 40M
Max Memory limit: 256M
Permalink Structure: /%postname%/
Language: en-US
Timezone: America/Denver
Debug Mode: Inactive

== Theme ==
Name: Hello Elementor
Version: 2.9.0
Author: Elementor Team
Child Theme: No

== User ==
Role: administrator
WP Profile lang: en_US
User Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.6 Safari/605.1.15

== Active Plugins ==
Code Snippets Pro
Version: 3.5.1
Author: Code Snippets Pro

Elementor
	Version: 3.17.1
	Author: Elementor.com

Elementor Pro
	Version: 3.16.2
	Author: Elementor.com

Image optimization service by Optimole
	Version: 3.10.0
	Author: Optimole
== Must-Use Plugins ==
Pantheon MU Plugin Loader
Version: 1.0
Author: Pantheon Systems

== Features ==
Custom Fonts: 0
Custom Icons: 0

== Integrations ==

== Elementor Experiments ==
Optimized DOM Output: Active
Improved Asset Loading: Active
Improved CSS Loading: Active
Inline Font Icons: Active
Additional Custom Breakpoints: Inactive
admin_menu_rearrangement: Inactive by default
Flexbox Container: Active
Upgrade Swiper Library: Active
Grid Container: Inactive
Default to New Theme Builder: Active by default
Hello Theme Header & Footer: Inactive
Editor Top Bar: Inactive by default
Optimized Gutenberg Loading: Inactive by default
Build with AI: Inactive by default
Landing Pages: Inactive
Nested Elements: Active
Lazy Load Background Images: Inactive
Optimize Image Loading: Inactive by default
Global Style Guide: Inactive
Page Transitions: Active by default
Notes: Active by default
Form Submissions: Active by default
Scroll Snap: Inactive
Menu: Inactive
Taxonomy Filter: Inactive

== Log ==

PHP: showing 4 of 4PHP: 2023-09-28 19:54:57 [notice X 1][../wp-content/plugins/elementor/core/upgrade/manager.php::70] version_compare(): Passing null to parameter https://github.com/elementor/elementor/pull/1 ($version1) of type string is deprecated [array (
'trace' => '
#0: Elementor\Core\Logger\Manager -> shutdown()
',
)]
PHP: 2023-09-28 20:00:39 [warning X 2][../wp-content/plugins/elementor/core/common/modules/ajax/module.php::175] Undefined array key "data" [array (
'trace' => '
#0: Elementor\Core\Logger\Manager -> shutdown()
',
)]
PHP: 2023-09-28 14:23:11 [error X 2][../wp-content/plugins/elementor-pro/modules/theme-builder/classes/conditions-manager.php::304] Uncaught Error: Call to a member function update_meta() on null in ../wp-content/plugins/elementor-pro/modules/theme-builder/classes/conditions-manager.php:304
Stack trace:
#0 ../wp-content/plugins/elementor-pro/modules/theme-builder/classes/conditions-manager.php(177): ElementorPro\Modules\ThemeBuilder\Classes\Conditions_Manager->save_conditions(11, Array)
https://github.com/elementor/elementor/pull/1 [internal function]: ElementorPro\Modules\ThemeBuilder\Classes\Conditions_Manager->ajax_save_theme_template_conditions(Array, Object(Elementor\Core\Common\Modules\Ajax\Module))
https://github.com/elementor/elementor/pull/2 ../wp-content/plugins/elementor/core/common/modules/ajax/module.php(175): call_user_func(Array, Array, Object(Elementor\Core\Common\Modules\Ajax\Module))
https://github.com/elementor/elementor/pull/3 ../wp-includes/class-wp-hook.php(310): Elementor\Core\Common\Modules\Ajax\Module->handle_ajax_request('')
https://github.com/elementor/elementor/pull/4 ../wp-includes/class-wp-hook.php(334): WP_Hook->apply_filters('', Array)
https://github.com/elementor/elementor/pull/5 ../wp-includes/plugin.php(517): WP_Hook->do_action(Array)
https://github.com/elementor/elementor/pull/6 ../wp-admin/admin-ajax.php(188): do_action('wp_ajax_element...')
https://github.com/elementor/elementor/pull/7 {main}
thrown [array (
'trace' => '
#0: Elementor\Core\Logger\Manager -> shutdown()
',
)]
PHP: 2023-10-20 03:30:34 [warning X 1][../wp-content/plugins/elementor/includes/elements/container.php::1847] Array to string conversion [array (
'trace' => '
#0: Elementor\Core\Logger\Manager -> shutdown()
',
)]

Log: showing 20 of 262023-10-09 09:56:56 [info] Elementor data updater process has been queued. [array (
'plugin' => 'Elementor',
'from' => '3.16.4',
'to' => '3.16.5',
)]
2023-10-09 09:57:11 [info] elementor-pro::elementor_pro_updater Started
2023-10-09 09:57:11 [info] Elementor Pro/Upgrades - _on_each_version Start
2023-10-09 09:57:11 [info] Elementor Pro/Upgrades - _on_each_version Finished
2023-10-09 09:57:11 [info] Elementor data updater process has been completed. [array (
'plugin' => 'Elementor Pro',
'from' => '3.14.1',
'to' => '3.16.2',
)]
2023-10-09 09:57:11 [info] Elementor data updater process has been queued. [array (
'plugin' => 'Elementor Pro',
'from' => '3.14.1',
'to' => '3.16.2',
)]
2023-10-18 11:32:30 [info] elementor::elementor_updater Started
2023-10-18 11:32:30 [info] Elementor/Upgrades - _on_each_version Start
2023-10-18 11:32:31 [info] Elementor/Upgrades - _on_each_version Finished
2023-10-18 11:32:31 [info] Elementor/Upgrades - _v_3_17_0_site_settings_updates Start
2023-10-18 11:32:31 [info] Elementor/Upgrades - _v_3_17_0_site_settings_updates Finished
2023-10-18 11:32:31 [info] Elementor data updater process has been completed. [array (
'plugin' => 'Elementor',
'from' => '3.16.5',
'to' => '3.16.6',
)]
2023-10-18 11:32:31 [info] Elementor data updater process has been queued. [array (
'plugin' => 'Elementor',
'from' => '3.16.5',
'to' => '3.16.6',
)]
2023-10-25 17:56:43 [info] elementor::elementor_updater Started
2023-10-25 17:56:43 [info] Elementor/Upgrades - _on_each_version Start
2023-10-25 17:56:43 [info] Elementor/Upgrades - _on_each_version Finished
2023-10-25 17:56:43 [info] Elementor/Upgrades - _v_3_17_0_site_settings_updates Start
2023-10-25 17:56:43 [info] Elementor/Upgrades - _v_3_17_0_site_settings_updates Finished
2023-10-25 17:56:43 [info] Elementor data updater process has been completed. [array (
'plugin' => 'Elementor',
'from' => '3.16.6',
'to' => '3.17.1',
)]
2023-10-25 17:56:43 [info] Elementor data updater process has been queued. [array (
'plugin' => 'Elementor',
'from' => '3.16.6',
'to' => '3.17.1',
)]

JS: showing 3 of 3JS: 2023-10-19 00:01:59 [error X 13][../wp-content/plugins/elementor/assets/js/frontend.min.js?ver=3.16.6:2:30893] elementor.settings.page is undefined
JS: 2023-10-19 16:06:45 [error X 3][../wp-admin/load-scripts.php?c=1&loadchunk_0=jquery-core,jquery-migrate,utils,moxiejs,plupload&ver=6.3.2:2:28722] Cannot read properties of undefined (reading 'value')
JS: 2023-10-26 02:03:52 [error X 3][../wp-admin/load-scripts.php?c=1&loadchunk_0=jquery-core,jquery-migrate,utils,moxiejs,plupload&ver=6.3.2:2:28729] undefined is not an object (evaluating 'l[0].value')

== Elementor - Compatibility Tag ==

Elementor Pro: Incompatible
== Elementor Pro - Compatibility Tag ==

````

</details>

**Comment by hein-obox** [comment]
Possible tweak: 
`--gap: var(--column-gap) var(--row-gap);`

Or:
`--gap: var(--column-gap) var(--row-gap, var(--column-gap));`

In that way, users have more control over the column and row custom styling.

**Comment by KingGrizzly** [comment]
@hein-obox Are you suggesting this as an override I can do with my CSS, or how Elementor should handle the software's CSS generation?

Here's the use case expanded. I'm basically begging anyone to watch this in full... @nicholaszein :-)

https://www.loom.com/share/de3ed18d15fc418e959014d384e1d7f0?sid=9317f2f6-c235-4d22-829f-be0029f2aa2f

This makes me wonder if Elementor could provide some kind of a gap size % slider alongside row and column gap for containers.

**Comment by hein-obox** [comment]
I meant it as an update to the Elementor plugin.

**Comment by KingGrizzly** [comment]
@hein-obox Thanks for clarifying. I agree. We really need column-gap and row-gap exposed individually — this would appear to support standard Flexbox CSS usage. Otherwise the "pencil" custom value field will be crippled for gaps. Placing variables and CSS in the custom value field will be undermined by the shorthand usage. Calc, for example, can't work — as demonstrated in my Loom video.

**Comment by KingGrizzly** [comment]
I'm basically begging Elementor to break these two values out and expose the column and row gaps separately — this seems easy. At the current moment it is impossible to vertically align certain layouts with either Flexbox or Grid.  Grid in Elementor is not feature rich enough yet (spans aren't supported) and Flexbox changes the width of child containers depending on the number of gaps, creating "columns" of unequal width. For example, these don't line up:

<img width="889" alt="misaligned" src="https://github.com/elementor/elementor/assets/48737505/62aedb64-117c-462a-9ef6-2a7283c7960d">

It's hard to sell clients on designs which can't be aligned in a fluid layout. Combine this with the current inability to import or cross-site paste ( https://github.com/elementor/elementor/issues/24111 ) and we are having some acute pain points.  This alignment issue can be fixed if the --gap-column for a given context were accessible to CSS calc and not combined with the --gap-row value.

**Comment by black-eye** [comment]
+1, Yes please!

**Comment by hmouhtar** [comment]
+1, really need this

**Comment by KingGrizzly** [comment]
Check this out. I believe these gap value behaviors are also related to a persistent bug relating to cross site pasting and export/import present since the same timeframe last October. It when Elementor went from a single gap value to a double I believe. The importer takes the first gap value and applies it to both column and row instead of maintaining two separate values.

https://www.loom.com/share/adda9c50442b4135b2ab10dec759d42b?sid=852315b8-0483-4716-b76a-b99487d6110e

**Comment by hein-obox** [comment]
@KingGrizzly @nicholaszein I have implemented this improvement. I will be added to version 3.26.

**Comment by hein-obox** [comment]
@KingGrizzly I really liked your suggestion.

The other issue with the copy and paste error will be handled in one of the following sprints.

**Comment by AF1QUE** [comment]
> @KingGrizzly I really liked your suggestion.
> 
> The other issue with the copy and paste error will be handled in one of the following sprints.

Hi @hein-obox, thank you so much for your attention on this, I would like to add there's another breaking bug related to copy-paste that doesn't copy over assets if they are in SVG format. As it's relevant to copy-paste, I am mentioning it here.

https://github.com/elementor/elementor/issues/21136

**Comment by nicholaszein** [comment]
Hey everyone!

We have great news! 🙌 

### 📢 We're happy to announce that this issue was resolved with the release of Elementor v3.26.0! 🥳 

You can now use the following CSS variables to target specific gap values for the Container element:

```css
selector {
    --row-gap: 2rem;
    --column-gap: 1rem;
}
```

The `--gap` variable is still present, and combines both the `--row-gap` and `--column-gap` variables.

✅ Feel free to check it out and update your plugin to the new version!

> _Check the changelog for more information:_
> [Changelog of Elementor and Elementor Pro](https://elementor.com/help/changelog-of-elementor/)

Cheers 🥂
